Improve traffic-managers podCIDR discovery

thallgren opened the pull request:
## Description

This PR contains two improvements to the pod CIDR discovery mechanism.
1. Performance improved by replacing the kates watchers with standard k8s APIs (#2084).
2. Make pod discovery strategy configurable in the Helm chart (#2023).

## Checklist

 - [x] I made sure to update `./CHANGELOG.md`.
 - [x] My change is adequately tested.

Commit:
Replace the kates package with a k8s SharedInformerFactory

The kates package provides nice and simple to use generalisations, but
it also introduces overhead and removes the ability to control some
aspects of watching. This commit replaces the traffic-manager's use of
the kates package with more standard k8s APIs.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix panic when traffic-manager reports info multiple times.

Fixes the
```
goroutine "/daemon/watch-cluster-info" exited with error: PANIC: close
of closed channel
```
that happens the second time the traffic-manager reports status info.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make traffic manager's strategy to find podCIDRs configurable.

Adds two environment variables to the traffic manager that controls
how it discovers what subnets the cluster is using for the pods. The
environment variables are configurable in the Helm chart.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

josecv commented on charts/telepresence/values.yaml:
```diff
@@ -70,6 +70,20 @@ grpc: {}
+#  nodePodCIDRs extract CIDRs from the podCIDR and podCIDRs field of the Node Spec.
+#  coverPodIPs  extract IPs from the podIP and podIPs field of the Pod Status and compute the CIDRs needed to cover those IPs.
+#  environment  use CIDRs listed in the space separated POD_CIDRS environment variable verbatim.
+#  auto         first tru nodePodCIDRs and if that fails, tru coverPodIPs
```
typo -- it says `tru` instead of `try`

josecv commented on cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go:
```diff
@@ -0,0 +1,249 @@
+	informer cache.SharedIndexInformer
+	subnets  []*net.IPNet
+	changed  time.Time
+	lock     sync.Mutex // Protects all access to ipsMap
```
should this read `Protects all access to subnets`?

josecv commented on cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go:
```diff
@@ -0,0 +1,249 @@
+	return changed
+}
+
+func (w *nodeWatcher) watch(ctx context.Context) {
```
maybe i'm just confused but it seems like watch is never invoked?

josecv commented on cmd/traffic/cmd/manager/internal/cluster/nodewatcher.go:
```diff
@@ -0,0 +1,249 @@
+	}
+}
+
+func (w *nodeWatcher) viable(ctx context.Context) bool {
+	nodes, err := w.lister.List(labels.Everything())
+	if err != nil {
+		dlog.Errorf(ctx, "unable to list nodes: %v", err)
+		return false
+	}
+
+	// Create the initial snapshot
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	changed := false
+	dlog.Infof(ctx, "Scanning %d nodes", len(nodes))
+	for _, node := range nodes {
+		if w.addLocked(nodeSubnets(ctx, node)) {
+			changed = true
+		}
+	}
+	if changed {
+		dlog.Infof(ctx, "Found %d subnets", len(w.subnets))
+		w.changed = time.Now()
+	} else {
+		dlog.Info(ctx, "No subnets found")
+	}
+	return changed
+}
```
it seems like this function combines initialization with a precondition check (i.e. "Are we actually able to watch subnets from nodes?") -- that's okay, but calling it twice in a row will make the second call return `false`, no? Maybe we should short-circuit return if `w.subnets` is already populated -- after all, if they're there then getting them is definitely viable

josecv commented on cmd/traffic/cmd/manager/internal/cluster/podwatcher.go:
```diff
@@ -0,0 +1,247 @@
+package cluster
```
same comments apply here as for the nodewatcher :)

josecv reviewed (commented):
looks good but i'm somewhat confused about the `watch` function --just wanna make sure that it actually _is_ being invoked, otherwise none of this will work 🙃

Commit:
Let constructor of nodeWatcher and podWatcher add the event handler.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Break set of net.IPNet into its own type.

Creates a new `subnet.Set` datatype with methods for adding, deleting,
and appending `net.IPNet` instances.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix bad comments.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Make viable() callable more than once.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix lint issues.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

thallgren commented:
The `watch` function was never called, but that only meant that future updates to podCIDRs after the initial scan where undetected. The initial scan happened anyway (in `viable`). Not good of course, but not a case of "none of this will work" either.

josecv commented on charts/telepresence/values.yaml:
```diff
@@ -79,7 +79,7 @@ podCIDRs: []
 #  coverPodIPs  extract IPs from the podIP and podIPs field of the Pod Status and compute the CIDRs needed to cover those IPs.
 #  environment  use CIDRs listed in the space separated POD_CIDRS environment variable verbatim.
-#  auto         first tru nodePodCIDRs and if that fails, tru coverPodIPs
+#  auto         first try nodePodCIDRs and if that fails, tru coverPodIPs
```
still says `tru coverPodIPs` :^)

Commit:
Fix test timeout when body.Read() returns error other than io.EOF

If the read at line 901 returned an error other than `io.EOF`, the
surrounding loop would continue in infinity. This commit ensures that
the loop brakes on any error.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Increase lint timeout to 3 minutes.

The 2 minutes timeout would sometimes fail.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>

Commit:
Fix weird root log checks in CLI test.

Signed-off-by: Thomas Hallgren <thomas@datawire.io>
